pkg/lib/size/size.go

```diff
@@ -1,26 +1,105 @@
 	"github.com/pkg/errors"
 )
 
 // Return the size in bytes of the provided endpoint. If the endpoint was archived, compressed or converted to
 // qcow2 the original image size is returned.
 func ImageSize(endpoint, accessKey, secKey string) (int64, error) {
-	ds, err := importer.NewDataStream(endpoint, accessKey, secKey)
+	ds, err := NewDataStream(endpoint, accessKey, secKey)
 	if err != nil {
 		return -1, errors.Wrapf(err, "unable to create data stream")
 	}
-	defer importer.CloseReaders(ds.Readers)
-	rdr := ds.Readers[len(ds.Readers)-1]
-	return Size(rdr, ds.Qemu)
+	defer ds.Close()
+	return Size(ds.Readers, ds.Qemu)
 }
 
 // Return the size of the endpoint corresponding to the passed-in reader.
-func Size(rdr io.ReadCloser, qemu bool) (int64, error) {
-	// TODO: figure out the size!
+func Size(readers []Reader, qemu bool) (int64, error) {
+	r := readers[len(readers)-1] // top-level reader
+	if qemu {
+		return qemuSize(r.Rdr)
+	}
+	switch r.RdrType {
+	case RdrFile:
+		return noSupport()
+	case RdrGz:
+		return noSupport()
+	case RdrHttp:
+		return httpSize(r.Rdr)
+	case RdrS3:
+		return noSupport()
+	case RdrTar:
+		return noSupport()
+	case RdrXz:
+		return noSupport()
+
+	default:
+		return 0, errors.Errorf("internal error: unsupported reader type %+v", r)
+	}
+}
+
+func noSupport() (int64, error) {
+	return -1, errors.New("unsupported format")
+}
+
+func httpSize(r io.Reader) (int64, error) {
+
 	return 0, nil
 }
+
+// Return the original (virtual) size of the file represented by the passed-in reader. This reader
+// is assumed to be a multi-reader, see importer.constructReaders(), and therefore a read loop is
+// necessary to handle partial reads.
+// Note: each read overwrites `buf` therefore the offset we "seek" in the file is variable.
+func qemuSize(r io.Reader) (int64, error) {
+	// format (defined: https://github.com/zchee/go-qcow2/blob/master/docs/specification.md#header)
+	// ...
+	// 24 - 31:   Virtual disk size in bytes
+	const (
+		sizeOffset = 24 // bytes
+		fieldSize  = 8  // bytes
+		bufSize    = sizeOffset + fieldSize
+	)
+
+	buf := make([]byte, bufSize)
+	var bufLen int
+
+	// the first multi-reader Read returns only 4 bytes (magic number) so a loop is needed
+	for {
+		len, err := r.Read(buf)
+		bufLen += len
+		if err == io.EOF { // unexpected
+			break
+		}
+		if err != nil {
+			return 0, errors.Wrapf(err, "cannot read qemu file")
+		}
+		if bufLen >= bufSize {
+			break
+		}
+	}
+	if bufLen < bufSize {
+		return 0, errors.Errorf("unexpected qemu buffer length: %d", bufLen)
+	}
+	delta := bufLen - bufSize // amount not read into buffer (== amount read by the multi reader)
+	offset := sizeOffset - delta
+	glog.V(Vdebug).Infof("qemuSize: reading at offset %d for %d bytes\n", offset, fieldSize)
+
+	// skip offset bytes and keep the "window" into buf `fieldSize` long
+	buf = buf[offset : bufSize-delta]
+	s := hex.EncodeToString(buf)
+	size, err := strconv.ParseInt(s, 16, 32)
```

Comment:
The function returns ( and the way the size is represented in the field) is 8 bytes.  32 should be 64 here.